### ISSUE TITLE
Invalid message signature error.

### ISSUE THREAD

**Issue by nighcoder** [clojupyter/clojupyter#122]
I'm opening this to document an issue I've found while working on feature/ipywidgets branch. It is not exclusive to the comms-atom layer, so many of the issues signalled here apply to the master branch as well.
The good news is that I've already tracked the problem and merged the fix in #120
## Summary
The message signature function in Clojupyter attempts to sign Clojure maps by casting them to strings, it fails when it attempts to sign a json-string parsed to a Clojure map that doesn't get cast to the identical json-string.
The best way to reproduce, is to find a deeply nested json (I've used a geo-json with lots of nested properties) and parse it to a Clojure data structure. You get a Clojure map with string keys. Now pass this map to `clojure.walk/keywordize-keywords` to get a Clojure map with keyword keys. Apply the json parser that Clojupyter use `cheshire.core/generate-string` to the two maps and compare the strings. If the generated strings are not identical, you found an incoming message that will fail the signature checking.

## The Error
Building widgets with deeply nested state can cause the kernel to crash and restart and to log the error:
> [E 22:35:48.491 Clojupyter] c8r.util-actions -- 
                                        java.lang.Thread.run              Thread.java:  748
          java.util.concurrent.ThreadPoolExecutor$Worker.run  ThreadPoolExecutor.java:  624
           java.util.concurrent.ThreadPoolExecutor.runWorker  ThreadPoolExecutor.java: 1149
                                                         ...                               
                           clojure.core.async/thread-call/fn                async.clj:  443
                                     clojupyter.zmq/start/fn                  zmq.clj:  108
                                      clojure.core.async/>!!                async.clj:  143
     clojure.core.async.impl.channels.ManyToManyChannel/put!             channels.clj:   83
                    clojure.core.async.impl.channels/chan/fn             channels.clj:  300
                                      clojure.core/map/fn/fn                 core.clj: 2742
                                                         ...                               
clojupyter.util-actions/wrap-report-and-absorb-exceptions/fn         util_actions.clj:  240
                                          clojure.core/apply                 core.clj:  665
                                                         ...                               
                                        clojure.core/comp/fn                 core.clj: 2569
         clojupyter.kernel.core/wrap-skip-shutdown-tokens/fn                 core.clj:   83
                                        clojure.core/comp/fn                 core.clj: 2569
                                     clojure.core/partial/fn                 core.clj: 2624
                                                         ...                               
                 clojure.spec.test.alpha/spec-checking-fn/fn                alpha.clj:  142
                                                         ...                               
                       clojupyter.messages/fn/frames->jupmsg             messages.clj:  289
java.lang.Exception: Invalid message signature
Exception in thread "async-thread-macro-11" java.lang.AssertionError: Assert failed: handle-request - internal error (':shell_port' neq '')
(= channel-port req-port)
        at clojupyter.kernel.handle_event_process$handle_event_or_channel_error.invokeStatic(handle_event_process.clj:23)
        at clojupyter.kernel.handle_event_process$handle_event_or_channel_error.invoke(handle_event_process.clj:16)
        at clojupyter.kernel.handle_event_process$run_dispatch_loop.invokeStatic(handle_event_process.clj:52)
        at clojupyter.kernel.handle_event_process$run_dispatch_loop.invoke(handle_event_process.clj:40)
        at clojupyter.kernel.handle_event_process$start_channel_thread$fn__28248.invoke(handle_event_process.clj:61)
        at clojure.core.async$thread_call$fn__5737.invoke(async.clj:443)
        at clojure.lang.AFn.run(AFn.java:22)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)

## Breaking it Down
The key error here is `java.lang.Exception: Invalid message signature`.
The function responsible with checking the signature is defined in `clojupyter.util/make-signer-checker`. It's defined as: 
``` 
(defn make-signer-checker
  [key]
  (let [mkchecker (fn mkchecker [signer]
                    (fn [{:keys [header parent-header metadata content preframes]}]
                      (let [payload-vec [header parent-header metadata content]
                            signature (.-signature preframes)
                            check-signature (signer payload-vec)]
                        (= check-signature (bytes->string* signature)))))
        signer	(if (empty? key)
                  (constantly "")
                  (fn signer [payload-vec]
                    (sha256-hmac (apply str (map json-str* payload-vec)) key)))]
    [signer (mkchecker signer)])) 
```
The issue is that the signer fn transforms the message before signing it by parsing it to a json string.
What I mean is, that a Clojure map `{:a 1 :b 2}` can be parsed to string in two valid ways `"{:a 1 :b 2}"` and `"{:b 2 :a 1}"` and while we generally don't care too much, the signing algorithm will generate different signatures for these two strings. Since the `checker` fn uses the `signer`, it will propagate the bug as well.

These two functions are used by the transducers for incoming and outgoing messages, both in `clojupyter.kernel.core`.
#### Inbound:
```
(defn- inbound-channel-transducer
  [port checker]
  (u!/wrap-report-and-absorb-exceptions
   (msgs/transducer-error port)
   (C (wrap-skip-shutdown-tokens (C (p msgs/frames->jupmsg checker)
                                    (p msgs/jupmsg->kernelreq port)))
      (logging-transducer (str "INBOUND:" port)))))
```
The function of interest here that gets called with the checker fn is `clojupyter.messages/frames->jupmsg`
```
(let [env-delim (vec (take 2 u/SEGMENT-ORDER))
      body-keys (vec (drop 2 u/SEGMENT-ORDER))
      _ (assert (= env-delim [:envelope :delimiter]))
      _ (assert (= body-keys [:signature :header :parent-header :metadata :content]))]

  (defn frames->jupmsg
    "Returns a map representing a Jupyter message.  `frames` must be a sequential collection of
  byte-arrays consisting of a "
    [check-signature frames]
    (let [delim u/IDSMSG-BYTES
          envelope (vec (take-while (complement u/delimiter-frame?) frames))
          _ (assert (< (count envelope) (count frames))) ;; we require a delimiter frame
          blobs (drop (inc (count envelope)) frames)     ;; we're dropping the delimiter too
          n-blobs (count body-keys)
          _ (assert (>= (count blobs) n-blobs))
          body (zipmap body-keys (take n-blobs blobs))
          signature (get body :signature) ; we want signature as bytes
          strbody (fmap u/bytes->string body)
          buffers (make-jupmsg-buffers (vec (drop n-blobs blobs)))
          preframes (make-jupmsg-preframes envelope delim signature)
          body-map (parse-json-and-build-jupmsg (select-keys strbody [:header :parent-header :metadata :content]))
          jupmsg (assoc body-map :preframes preframes :buffers buffers)]
      (when-not (check-signature jupmsg)
        (let [msg "Invalid message signature"]
          (log/error msg)
         (throw (Exception. msg))))
      (s/assert ::jsp/jupmsg jupmsg))))
```
There's a few happening here, so we'll take it one by one. We'll focus on the inner let form:
* `body` gets assigned to a map of byte-array values.
* we extract the `signature` from the `body` and build the `preframes` and `buffers`.
* we parse the body map to strbody to get a map of string values in `strbody`.
* we parse the string and coerce the values to clojure maps in `body-map`.
* we build the full jupyter message by adding the `:preframes` and `:buffers` to `body-map`.
* we pass the jupmsg to signature checker fn to validate.

There's two issues with this timeline:
1. We attempt to parse a string before knowing if the signature is valid or not.
2. The checker function, expects clojure maps in order to use the existing signer function, which can't reliably sign clojure maps.

The `checker` function should accept byte-arrays or strings, since those generate unique signatures. It makes sense to use the json strings for this, since we use them when building and parsing the jupyter frames.

#### Outbound:
```
(defn- outbound-channel-transducer
  [port signer]
  (u!/wrap-report-and-absorb-exceptions
   (msgs/transducer-error port)
   (C (wrap-skip-shutdown-tokens
       (C (fn [krsp] (extract-kernel-response-byte-arrays krsp))
          (fn [krsp] (replace-comm-atoms-with-references krsp))
          (fn [krsp] (msgs/kernelrsp->jupmsg port signer krsp))))
      (logging-transducer (str "OUTBOUND:" port))
      (wrap-skip-shutdown-tokens (fn [jupmsg] (msgs/jupmsg->frames jupmsg))))))
```
Here the implementation poses a challenge, because it expects to sign the message at `kernelrsp->jupmsg`, but builds the json-strings in the last step by `jupmsg->frames`.
```
(defn kernelrsp->jupmsg
  ([port signer kernel-rsp]
   (kernelrsp->jupmsg port signer kernel-rsp {}))
  ([port signer
    {:keys [rsp-content rsp-msgtype rsp-socket rsp-metadata rsp-buffers req-message]}
    {:keys [messageid now] :as opts}]
   (let [messageid	(str (or messageid (u!/uuid)))
         now		(or now (u!/now))
         session-id	(message-session req-message)
         username	(message-username req-message)
         header		(make-jupmsg-header messageid rsp-msgtype username session-id now PROTOCOL-VERSION)
         parent-header	(message-header req-message)
         metadata	(or rsp-metadata {})
         rsp-buffers	(or rsp-buffers [])
         envelope	(if (= rsp-socket :iopub_port)
                          [(u/get-bytes rsp-msgtype)]
                          (message-envelope req-message))
         signature	(u/get-bytes (signer [header parent-header metadata rsp-content]))]
     (make-jupmsg envelope signature header parent-header metadata rsp-content rsp-buffers))))
```
The thing to note here is that both kernel response and jupyter messsage are Clojure maps. These maps are used to generate the signature and pass it along in a map to `jupmsh->frames`.
```
(let [;; don't want envelope+delimiter in sig scope:
      jupmsg-keys (drop 2 u/SEGMENT-ORDER)]

  (defn jupmsg->frames
    [{:keys [header parent-header metadata content preframes buffers] :as jupmsg}]
    (assert (every? (complement nil?) (vals (select-keys jupmsg jupmsg-keys))))
    (let [envelope	(.-envelope preframes)
          signature	(.-signature preframes)
          _		(assert envelope)
          _		(assert signature)
          payload-vec	(mapv u/json-str [header parent-header metadata content])
          byte-buffers	(when buffers
                          (.-buffers buffers))]
        (assert (s/valid? ::sp/byte-arrays envelope))
        (->> (concat envelope
                     [u/IDSMSG-BYTES (u/get-bytes signature)]
                     (mapv u/get-bytes payload-vec)
                     byte-buffers)
             vec
             (s/assert ::msp/frames)))))
```
Here the map is transformed to the right format to be passed along the channel. Note that `payload-vec` contains the json string we'll pass along to front-end and also the string we'll like to sign.
Unfortunately, the signature was already generated using a different call to json-string in the `signer` function.

## The Proposed Fix
* We change `signer` to work on json strings. 
* We change `frames->jupmsg` to call the checker with a map of json strings. We parse the string only when the signature is valid. There's no need to parse junk from the wire.
* We change the outbound logic to generate and attach the signature in the last step, at `jupmsg->frames`. `kernelrsp->jupmsg` will no longer attempt to generate signatures on maps. In #120 I've done that by changing the fn signature of `jupmsg->frames` to accept the `signer` fn as extra argument and removed the `signer` fn from the `kernelrsp->jupmsg` argument list. The patch passes along an empty byte-array as signature since many functions that build the jupyter message rely on finding a singnature in the `:preframes` field.
* This is not exactly a fix, but a design feature. Remember I said the first thing that happened whas that the kernel restarted and the error printed in the log. The kernel should not crash on invalid messages, even if the signature does not match. It may receive corrupt messages from a bad network connection or an attacker sending junk on the wire. If it fails to process any of those, it acts as *snowflake* software. 
To prevent this brittle behaviour, I've also changed the mssage dispatch loop to continue on error and log the errors.

**Comment by krukow** [comment]
@slavad @nighcoder I think https://github.com/clojupyter/clojupyter/pull/135 should fix this. Would you please review that solution relative to https://github.com/clojupyter/clojupyter/pull/120

**Comment by nighcoder** [comment]
@krukow I've checked your PR relative to the changes I've made to tackle the same issue and here are my thoughts.
For reference, the changes I'm talking about are added in commit 12de06260ff2b4bc7f3f060041b5aeab3bf9c68d

I'll reiterate the list of proposed fixes against #135:

> * We change signer to work on json strings.

This one is fixed in #135 


> * We change frames->jupmsg to call the checker with a map of json strings.

The first part of this point is fixed in #135, but the second part is no longer true:

> We parse the string only when the signature is valid. There's no need to parse junk from the wire.

As it stands, we attempt to parse the message in `frames->jupmsg` (line 284 in messages.clj) before checking if the signature is correct.


> * We change the outbound logic to generate and attach the signature in the last step, at jupmsg->frames. kernelrsp->jupmsg will no longer attempt to generate signatures on maps. In Feature/ipywidgets #120 I've done that by changing the fn signature of jupmsg->frames to accept the signer fn as extra argument and removed the signer fn from the kernelrsp->jupmsg argument list. The patch passes along an empty byte-array as signature since many functions that build the jupyter message rely on finding a singnature in the :preframes field.

Nothing is changed in #135 regarding this issue. The kernel still attempts to sign the message before serializing it before hand. Check `jupmsg->frames` where we generate the vector of strings that we put on the wire (line 307 in messages.clj). At that point we rely on the signature being present in the preframes and the signature is computed in `jupmsg->frames` (line 352). As you see, we serialize the our message, which is a vector of maps at two different points and there's no guarantee that the resulting string will be equal.

> The kernel should not crash on invalid messages, even if the signature does not match.

I still stand by that statement. I think crashing and exiting is an exaggerated response to an message with invalid signature.

**Comment by krukow** [comment]
Hey @nighcoder - I agree with your assessment here.

Given the relative mess of branches at the moment may I propose that we break things down a bit and incrementally get to the desired state.

The plan might be:

(i) test, review and merge https://github.com/clojupyter/clojupyter/pull/128
(i) you open a PR to make the additional robustness improvements (as you mention above) - or alternatively if you don't have time I can take a stab at that over the weekend (or next :) ) based on  https://github.com/clojupyter/clojupyter/issues/122#issuecomment-911058837 

Let's keep this issue open until all propsed improvements are done.

**Comment by nighcoder** [comment]
@krukow, I can make a PR with the changes, it just depends on what order you plan to merge them and in what branch.

**Comment by nighcoder** [comment]
Fixed in 936c73f3b83787c2ab9efc0d0a7a64ef3ea8f441.
